service/history/shard/context.go

```diff
@@ -252,6 +260,69 @@ func (s *contextImpl) UpdateTransferClusterAckLevel(cluster string, ackLevel int
 	return s.updateShardInfoLocked()
 }
 
+func (s *contextImpl) GetTransferProcessingQueueStates(cluster string) []*history.ProcessingQueueState {
+	s.RLock()
+	defer s.RUnlock()
+
+	// if we can find corresponding processing queue states
+	if states, ok := s.transferProcessingQueueStates.StatesByCluster[cluster]; ok {
+		return states
+	}
+
+	// check if we can find corresponding ack level
+	var ackLevel int64
+	var ok bool
+	if ackLevel, ok = s.shardInfo.ClusterTransferAckLevel[cluster]; !ok {
+		// otherwise, default to existing ack level, which belongs to local cluster
+		// this can happen if you add more cluster
+		ackLevel = s.shardInfo.TransferAckLevel
+	}
+
+	// otherwise, create default queue state based on existing ack level,
+	// which belongs to local cluster. this can happen if you add more cluster
+	return []*history.ProcessingQueueState{
+		{
+			Level:    common.Int32Ptr(0),
+			AckLevel: common.Int64Ptr(ackLevel),
+			MaxLevel: common.Int64Ptr(math.MaxInt64),
+			DomainFilter: &history.DomainFilter{
+				ReverseMatch: common.BoolPtr(true),
+			},
+		},
+	}
+}
+
+func (s *contextImpl) UpdateTransferProcessingQueueStates(cluster string, states []*history.ProcessingQueueState) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(states) == 0 {
+		return errors.New("Empty transfer processing queue states")
+	}
+
+	if s.transferProcessingQueueStates.StatesByCluster == nil {
```

Comment:
when it will be nil?

Reply:
It should not be nil, but adding the check for safety. I can remove if you have strong opinion.